File: src/assembler/skills-assembler.ts
/**
 * SkillsAssembler — assembles .claude/skills/ from templates based on project config.
 *
 * Migrated from Python `assembler/skills.py` (285 lines).
 * Selection logic lives in {@link ./skills-selection.ts}; this module handles file I/O.
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import {
  copyTemplateTree,
  copyTemplateTreeIfExists,
  copyTemplateFile,
} from "./copy-helpers.js";
import { buildInfraPackRules } from "../domain/skill-registry.js";
import { getStackPackName } from "../domain/stack-pack-mapping.js";
import {
  selectConditionalSkills,
  selectKnowledgePacks,
} from "./skills-selection.js";

const SKILLS_TEMPLATES_DIR = "skills-templates";
const CORE_DIR = "core";
const CONDITIONAL_DIR = "conditional";
const KNOWLEDGE_PACKS_DIR = "knowledge-packs";
const INFRA_PATTERNS_DIR = "infra-patterns";
const STACK_PATTERNS_DIR = "stack-patterns";
const LIB_DIR = "lib";
const SKILL_MD = "SKILL.md";
const SKILLS_OUTPUT = "skills";

/** Assembles skills from templates based on project config. */
export class SkillsAssembler {
  /** Scan core skills directories, returning skill names. */
  selectCoreSkills(resourcesDir: string): string[] {
    const corePath = path.join(
      resourcesDir, SKILLS_TEMPLATES_DIR, CORE_DIR,
    );
    if (!fs.existsSync(corePath)) return [];
    const skills: string[] = [];
    const entries = fs.readdirSync(corePath, { withFileTypes: true });
    for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
      if (!entry.isDirectory()) continue;
      if (entry.name === LIB_DIR) {
        const libPath = path.join(corePath, LIB_DIR);
        const subs = fs.readdirSync(libPath, { withFileTypes: true });
        for (const sub of subs.sort((a, b) => a.name.localeCompare(b.name))) {
          if (sub.isDirectory()) {
            skills.push(`${LIB_DIR}/${sub.name}`);
          }
        }
      } else {
        skills.push(entry.name);
      }
    }
    return skills;
  }

  /** Evaluate feature gates and return conditional skill names. */
  selectConditionalSkills(config: ProjectConfig): string[] {
    return selectConditionalSkills(config);
  }

  /** Select knowledge packs based on config. */
  selectKnowledgePacks(config: ProjectConfig): string[] {
    return selectKnowledgePacks(config);
  }

  /** Main entry point: assemble all skills. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
  ): string[] {
    const results: string[] = [];
    results.push(...this.assembleCore(resourcesDir, outputDir, engine));
    results.push(
      ...this.assembleConditional(config, resourcesDir, outputDir, engine),
    );
    results.push(
      ...this.assembleKnowledge(config, resourcesDir, outputDir, engine),
    );
    return results;
  }

  private copyCoreSkill(
    skillName: string, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string {
    const src = path.join(
      resourcesDir, SKILLS_TEMPLATES_DIR, CORE_DIR, skillName,
    );
    const dest = path.join(outputDir, SKILLS_OUTPUT, skillName);
    return copyTemplateTree(src, dest, engine);
  }

  private copyConditionalSkill(
    skillName: string, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string | null {
    const src = path.join(
      resourcesDir, SKILLS_TEMPLATES_DIR, CONDITIONAL_DIR, skillName,
    );
    const dest = path.join(outputDir, SKILLS_OUTPUT, skillName);
    return copyTemplateTreeIfExists(src, dest, engine);
  }

  private copyKnowledgePack(
    packName: string, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string | null {
    const src = path.join(
      resourcesDir, SKILLS_TEMPLATES_DIR, KNOWLEDGE_PACKS_DIR, packName,
    );
    if (!fs.existsSync(src)) return null;
    const dest = path.join(outputDir, SKILLS_OUTPUT, packName);
    fs.mkdirSync(dest, { recursive: true });
    const skillMdSrc = path.join(src, SKILL_MD);
    if (fs.existsSync(skillMdSrc)) {
      copyTemplateFile(skillMdSrc, path.join(dest, SKILL_MD), engine);
    }
    this.copyNonSkillItems(src, dest);
    return dest;
  }

  private copyNonSkillItems(src: string, dest: string): void {
    const entries = fs.readdirSync(src, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name === SKILL_MD) continue;
      const target = path.join(dest, entry.name);
      if (fs.existsSync(target)) continue;
      const source = path.join(src, entry.name);
      if (entry.isDirectory()) {
        fs.cpSync(source, target, { recursive: true });
      } else {
        fs.copyFileSync(source, target);
      }
    }
  }

  private copyStackPatterns(
    config: ProjectConfig, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string | null {
    const packName = getStackPackName(config.framework.name);
    if (!packName) return null;
    const src = path.join(
      resourcesDir, SKILLS_TEMPLATES_DIR,
      KNOWLEDGE_PACKS_DIR, STACK_PATTERNS_DIR, packName,
    );
    const dest = path.join(outputDir, SKILLS_OUTPUT, packName);
    return copyTemplateTreeIfExists(src, dest, engine);
  }

  private copyInfraPatterns(
    config: ProjectConfig, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string[] {
    const results: string[] = [];
    for (const [packName, condition] of buildInfraPackRules(config)) {
      if (!condition) continue;
      const src = path.join(
        resourcesDir, SKILLS_TEMPLATES_DIR,
        KNOWLEDGE_PACKS_DIR, INFRA_PATTERNS_DIR, packName,
      );
      const dest = path.join(outputDir, SKILLS_OUTPUT, packName);
      const copied = copyTemplateTreeIfExists(src, dest, engine);
      if (copied !== null) results.push(copied);
    }
    return results;
  }

  private assembleCore(
    resourcesDir: string, outputDir: string, engine: TemplateEngine,
  ): string[] {
    return this.selectCoreSkills(resourcesDir).map(
      (skill) => this.copyCoreSkill(skill, resourcesDir, outputDir, engine),
    );
  }

  private assembleConditional(
    config: ProjectConfig, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string[] {
    const results: string[] = [];
    for (const skill of this.selectConditionalSkills(config)) {
      const copied = this.copyConditionalSkill(
        skill, resourcesDir, outputDir, engine,
      );
      if (copied !== null) results.push(copied);
    }
    return results;
  }

  private assembleKnowledge(
    config: ProjectConfig, resourcesDir: string,
    outputDir: string, engine: TemplateEngine,
  ): string[] {
    const results: string[] = [];
    for (const pack of this.selectKnowledgePacks(config)) {
      const copied = this.copyKnowledgePack(
        pack, resourcesDir, outputDir, engine,
      );
      if (copied !== null) results.push(copied);
    }
    const stack = this.copyStackPatterns(
      config, resourcesDir, outputDir, engine,
    );
    if (stack !== null) results.push(stack);
    results.push(
      ...this.copyInfraPatterns(config, resourcesDir, outputDir, engine),
    );
    return results;
  }
}

File: java/src/main/java/dev/iadev/application/assembler/SkillsAssembler.java
package dev.iadev.application.assembler;

import dev.iadev.config.ContextBuilder;
import dev.iadev.domain.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;

import dev.iadev.domain.model.ContextBudget;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Assembles {@code .claude/skills/} from templates based on
 * project configuration.
 *
 * <p>This is the second assembler in the pipeline (position 2
 * of 23 per RULE-005). It generates three categories of
 * skills:
 * <ol>
 *   <li>Core skills — always included regardless of
 *       profile</li>
 *   <li>Conditional skills — included based on feature
 *       gates evaluated by {@link SkillsSelection}</li>
 *   <li>Knowledge packs — always included, plus
 *       stack-specific and infrastructure patterns</li>
 * </ol>
 *
 * <p>Policy and path-resolution responsibilities are
 * delegated to cohesive helpers (audit M-002):</p>
 * <ul>
 *   <li>{@link ProtectedNamePolicy} — owns the reserved
 *       top-level names set used by the prune pass.</li>
 *   <li>{@link SkillCategoryResolver} — owns skill-name to
 *       source-path resolution across the EPIC-0036
 *       category-aware layout (flat output, hierarchical
 *       source).</li>
 * </ul>
 *
 * @see Assembler
 * @see SkillsSelection
 * @see SkillsCopyHelper
 * @see ProtectedNamePolicy
 * @see SkillCategoryResolver
 */
public final class SkillsAssembler implements Assembler {

    private static final String SKILLS_TEMPLATES_DIR =
            "targets/claude/skills";
    private static final String CORE_DIR = "core";
    private static final String CONDITIONAL_DIR =
            "conditional";
    private static final String SHARED_DIR = "_shared";
    private static final String SKILLS_OUTPUT = "skills";

    private final Path resourcesDir;

    /**
     * Creates a SkillsAssembler using classpath resources.
     */
    public SkillsAssembler() {
        this(resolveClasspathResources());
    }

    /**
     * Creates a SkillsAssembler with an explicit resources
     * directory.
     *
     * @param resourcesDir the base resources directory
     */
    public SkillsAssembler(Path resourcesDir) {
        this.resourcesDir = resourcesDir;
    }

    /** {@inheritDoc} */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        java.util.Map<String, Object> context =
                ContextBuilder.buildContext(config);
        List<String> generated = new ArrayList<>();
        generated.addAll(
                assembleCore(outputDir, engine, context));
        generated.addAll(
                assembleConditional(
                        config, outputDir, engine, context));
        // EPIC-0051 (ADR-0013): knowledge packs moved to
        // .claude/knowledge/ under KnowledgeAssembler; no
        // longer emitted by SkillsAssembler.
        assembleShared(outputDir)
                .ifPresent(generated::add);
        pruneStaleSkills(outputDir, generated);
        return generated;
    }

    /**
     * Copies the source-of-truth {@code _shared/} directory
     * (peer of {@code core/} / {@code conditional/} /
     * {@code knowledge-packs/}) to
     * {@code outputDir/skills/_shared/} verbatim.
     *
     * <p>Story-0047-0001 introduces {@code _shared/} for
     * cross-cutting Markdown snippets referenced by
     * consumer skills via Markdown relative links
     * (per ADR-0011 — shared-snippets inclusion strategy).
     * For links like {@code ../_shared/error-handling.md}
     * in a generated {@code skills/x-git-commit/SKILL.md}
     * to resolve at runtime, the target directory must ship
     * in the same output tree.</p>
     *
     * <p>When the source {@code _shared/} directory does not
     * exist, this method is a no-op and returns
     * {@link Optional#empty()}.</p>
     *
     * <p>Returning the output path (when present) lets
     * {@link #pruneStaleSkills(Path, List)} keep the
     * directory across reruns.</p>
     *
     * @param outputDir the output root directory
     * @return the absolute output path of the copied
     *         {@code _shared/} directory, or empty when the
     *         source directory is absent
     */
    private Optional<String> assembleShared(Path outputDir) {
        Path sharedSrc = sharedPath();
        if (!Files.exists(sharedSrc)
                || !Files.isDirectory(sharedSrc)) {
            return Optional.empty();
        }
        Path dest = outputDir.resolve(
                SKILLS_OUTPUT + "/" + SHARED_DIR);
        CopyHelpers.copyDirectory(sharedSrc, dest);
        return Optional.of(dest.toString());
    }

    private Path sharedPath() {
        return resourcesDir.resolve(
                SKILLS_TEMPLATES_DIR + "/" + SHARED_DIR);
    }

    /**
     * Removes top-level directories under
     * {@code outputDir/skills/} that are not present in
     * {@code generatedPaths} and are not reserved by
     * {@link ProtectedNamePolicy}.
     *
     * <p>Without this pass the output is additive-only:
     * skills renamed or removed in the source of truth
     * persist indefinitely as stale copies. Stray files
     * at the skills root are never touched — only
     * directories are candidates for removal.</p>
     *
     * @param outputDir       the root output directory
     * @param generatedPaths  absolute paths of every skill
     *                        written in this run
     */
    void pruneStaleSkills(
            Path outputDir, List<String> generatedPaths) {
        Path skillsDir = outputDir.resolve(SKILLS_OUTPUT);
        if (!Files.isDirectory(skillsDir)) {
            return;
        }
        Set<String> expected = expectedTopLevelNames(
                generatedPaths, skillsDir);
        try (Stream<Path> stream = Files.list(skillsDir)) {
            stream
                    .filter(Files::isDirectory)
                    .filter(p -> !isRetained(p, expected))
                    .forEach(SkillsAssembler::deleteStrictly);
        } catch (IOException e) {
            throw new UncheckedIOException(
                    "Failed to prune stale skills in: %s"
                            .formatted(skillsDir), e);
        }
    }

    /**
     * Attempts to delete {@code path} recursively and raises
     * an {@link UncheckedIOException} if the directory is still
     * present afterwards.
     *
     * <p>Wraps {@code CopyHelpers.deleteQuietly} so that silent
     * failures (insufficient permissions, open file handles on
     * Windows, EBUSY, etc.) do not leave a half-pruned output.
     * The stale directory surviving is a contract violation of
     * the destructive prune and must surface to the operator.</p>
     *
     * @param path the directory to delete
     * @throws UncheckedIOException if {@code path} exists after
     *                              the delete attempt
     */
    private static void deleteStrictly(Path path) {
        try {
            Files.walkFileTree(path, new SimpleFileVisitor<>() {
                @Override
                public FileVisitResult visitFile(
                        Path file, BasicFileAttributes attrs)
                        throws IOException {
                    Files.delete(file);
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult postVisitDirectory(
                        Path dir, IOException exc)
                        throws IOException {
                    if (exc != null) {
                        throw exc;
                    }
                    Files.delete(dir);
                    return FileVisitResult.CONTINUE;
                }
            });
        } catch (IOException e) {
            throw new UncheckedIOException(
                    "Failed to prune stale skill directory: "
                            + path,
                    e);
        }
    }

    private static boolean isRetained(
            Path dir, Set<String> expected) {
        String name = dir.getFileName().toString();
        return ProtectedNamePolicy.isProtected(name)
                || expected.contains(name);
    }

    private static Set<String> expectedTopLevelNames(
            List<String> generatedPaths, Path skillsDir) {
        Path root = skillsDir.toAbsolutePath().normalize();
        return generatedPaths.stream()
                .map(Path::of)
                .map(Path::toAbsolutePath)
                .map(Path::normalize)
                .filter(p -> p.startsWith(root)
                        && !p.equals(root))
                .map(p -> root.relativize(p)
                        .getName(0).toString())
                .collect(Collectors.toSet());
    }

    /**
     * Scans the core skills directory and returns skill names.
     *
     * <p>Delegates to {@link SkillCategoryResolver} so the
     * hierarchical-source / flat-output mapping is owned by a
     * single helper. See
     * {@link SkillCategoryResolver#listSkills(Path)} for scan
     * order and sort contract.</p>
     *
     * @return list of core skill names in scan order
     */
    List<String> selectCoreSkills() {
        return SkillCategoryResolver.listSkills(corePath());
    }

    private Path corePath() {
        return resourcesDir.resolve(
                SKILLS_TEMPLATES_DIR + "/" + CORE_DIR);
    }

    private Path conditionalPath() {
        return resourcesDir.resolve(
                SKILLS_TEMPLATES_DIR + "/"
                        + CONDITIONAL_DIR);
    }

    private List<String> assembleCore(
            Path outputDir,
            TemplateEngine engine,
            java.util.Map<String, Object> context) {
        List<String> generated = new ArrayList<>();
        for (String skill : selectCoreSkills()) {
            String result = copyCoreSkill(
                    skill, outputDir, engine, context);
            generated.add(result);
        }
        return generated;
    }

    private List<String> assembleConditional(
            ProjectConfig config,
            Path outputDir,
            TemplateEngine engine,
            java.util.Map<String, Object> context) {
        List<String> generated = new ArrayList<>();
        List<String> conditional =
                SkillsSelection.selectConditionalSkills(
                        config);
        for (String skill : conditional) {
            copyConditionalSkill(
                    skill, outputDir, engine, context)
                    .ifPresent(generated::add);
        }
        return generated;
    }

    private String copyCoreSkill(
            String skillName,
            Path outputDir,
            TemplateEngine engine,
            java.util.Map<String, Object> context) {
        Path src = SkillCategoryResolver.resolveSourcePath(
                corePath(), skillName);
        Path dest = outputDir.resolve(
                SKILLS_OUTPUT + "/" + skillName);
        CopyHelpers.copyDirectory(src, dest);
        CopyHelpers.replacePlaceholdersInDir(
                dest, engine, context);
        injectBudgetField(src, dest);
        return dest.toString();
    }

    private void injectBudgetField(
            Path srcDir, Path destDir) {
        Path destSkill = destDir.resolve("SKILL.md");
        if (!Files.exists(destSkill)) {
            return;
        }
        String content = CopyHelpers.readFile(destSkill);
        int lineCount =
                (int) content.lines().count();
        ContextBudget budget =
                ContextBudget.fromLineCount(lineCount);
        String injected =
                FrontmatterInjector.injectContextBudget(
                        content, budget);
        CopyHelpers.writeFile(destSkill, injected);
    }


    private Optional<String> copyConditionalSkill(
            String skillName,
            Path outputDir,
            TemplateEngine engine,
            java.util.Map<String, Object> context) {
        Path src = SkillCategoryResolver.resolveSourcePath(
                conditionalPath(), skillName);
        if (!Files.exists(src)
                || !Files.isDirectory(src)) {
            return Optional.empty();
        }
        Path dest = outputDir.resolve(
                SKILLS_OUTPUT + "/" + skillName);
        CopyHelpers.copyDirectory(src, dest);
        CopyHelpers.replacePlaceholdersInDir(
                dest, engine, context);
        injectBudgetField(src, dest);
        return Optional.of(dest.toString());
    }

    private static Path resolveClasspathResources() {
        return dev.iadev.util.ResourceResolver
                .resolveResourceDir("shared")
                .getParent();
    }
}
